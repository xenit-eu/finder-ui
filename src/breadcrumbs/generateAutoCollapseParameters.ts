import invariant from "tiny-invariant";

export type GenerateAutoCollapseParameters_t = {
    itemsBeforeCollapse?: number | { min?: number, max?: number },
    itemsAfterCollapse?: number | { min?: number, max?: number },
    maxItems?: number,
    size: number,
};

type AutoCollapseOutput_t = {
    maxItems: number,
    itemsBeforeCollapse: number,
    itemsAfterCollapse: number,
};

export type GenerateAutoCollapseParametersOutput_t = {
    priority: number,
    output: AutoCollapseOutput_t,
};
function sumIfDefined(a: number | undefined, b: number | undefined): number | undefined {
    if (a !== undefined && b !== undefined) {
        return a + b;
    }
    return undefined;
}

function getMin(n: number | { min?: number } | undefined): number | undefined {
    if (typeof n !== "object") {
        return n;
    }
    return n.min;
}

function getMax(n: number | { max?: number } | undefined): number | undefined {
    if (typeof n !== "object") {
        return n;
    }
    return n.max;
}

function* autoCollapseParametersGen(parameters: GenerateAutoCollapseParameters_t): Generator<GenerateAutoCollapseParametersOutput_t> {
    const minItemsBeforeCollapse = getMin(parameters.itemsBeforeCollapse) || 1;
    const minItemsAfterCollapse = getMin(parameters.itemsAfterCollapse) || 1;
    // If max items is set, only one option is available, that with the configured options
    if (parameters.maxItems !== undefined) {
        yield {
            priority: 1,
            output: {
                maxItems: parameters.maxItems,
                itemsBeforeCollapse: minItemsBeforeCollapse,
                itemsAfterCollapse: minItemsAfterCollapse,
            },
        };
        return;
    }
    // Showing the breadcrumbs full size is preferred
    yield {
        priority: parameters.size + 1,
        output: {
            maxItems: Math.max(parameters.size, minItemsAfterCollapse + minItemsAfterCollapse + 1),
            itemsBeforeCollapse: minItemsBeforeCollapse,
            itemsAfterCollapse: minItemsAfterCollapse,
        },
    };

    const maxItemsBeforeCollapse = getMax(parameters.itemsBeforeCollapse);
    const maxItemsAfterCollapse = getMax(parameters.itemsAfterCollapse);

    // Calculate minimum and maximum number of items that are shown
    const minItemsShown = minItemsBeforeCollapse + minItemsAfterCollapse;
    const maxItemsShown = sumIfDefined(maxItemsBeforeCollapse, maxItemsAfterCollapse) || parameters.size - 1;

    let itemsBeforeCollapse = getMin(parameters.itemsBeforeCollapse) || 1;
    let itemsAfterCollapse = getMin(parameters.itemsAfterCollapse) || 1;

    // Step through the options, incrementing the collapse item that is furthest away from its maximum value with every step
    for (let i = minItemsShown; i <= maxItemsShown; i++) {
        if (itemsBeforeCollapse + itemsAfterCollapse >= parameters.size - 1) {
            break;
        }
        yield {
            priority: i,
            output: {
                maxItems: parameters.size - 1,
                itemsBeforeCollapse,
                itemsAfterCollapse,
            },
        };
        const beforeCollapsePref = calculatePreference(minItemsBeforeCollapse, maxItemsBeforeCollapse || parameters.size, itemsBeforeCollapse);
        const afterCollapsePref = calculatePreference(minItemsAfterCollapse, maxItemsAfterCollapse || parameters.size, itemsAfterCollapse);
        if (beforeCollapsePref < afterCollapsePref) {
            itemsAfterCollapse++;
        } else {
            itemsBeforeCollapse++;
        }
    }

}

/**
 * Calculates where between minValue & maxValue the currentValue lays, as a fraction
 */
function calculatePreference(minValue: number, maxValue: number, currentValue: number): number {
    if (currentValue >= maxValue) {
        return -1;
    }
    invariant(minValue <= maxValue);
    invariant(minValue <= currentValue);
    invariant(currentValue <= maxValue);
    // Normalize to minValue = 0
    maxValue = maxValue - minValue;
    currentValue = currentValue - minValue;

    return 1 - currentValue / maxValue;
}

function toArray<T>(generator: Generator<T>): readonly T[] {
    const data: T[] = [];
    let value = generator.next();

    while (!value.done) {
        data.push(value.value);
        value = generator.next();
    }

    return data;
}

export default function generateAutoCollapseParameters(parameters: GenerateAutoCollapseParameters_t): readonly GenerateAutoCollapseParametersOutput_t[] {
    return toArray(autoCollapseParametersGen(parameters));
}
