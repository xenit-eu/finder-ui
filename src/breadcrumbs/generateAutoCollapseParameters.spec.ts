import generateAutoCollapseParameters, { GenerateAutoCollapseParametersOutput_t } from "./generateAutoCollapseParameters";

function sortByPriority(params: readonly GenerateAutoCollapseParametersOutput_t[]): readonly GenerateAutoCollapseParametersOutput_t[] {
    return params.slice().sort((a, b) => b.priority - a.priority);
}

describe("breadcrumbs/generateAutoCollapseParameters", () => {
    it("generates one set of parameters when maxItems is set", () => {
        const data = generateAutoCollapseParameters({
            size: 10,
            maxItems: 4,
        });

        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when no constraints are set", () => {
        const data = generateAutoCollapseParameters({
            size: 10,
        });

        expect(data).toContainEqual({
            priority: 11,
            output: {
                maxItems: 10,
                itemsBeforeCollapse: 1,
                itemsAfterCollapse: 1,
            },
        });
        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having itemsBeforeCollapse fixed", () => {
        const data = generateAutoCollapseParameters({
            size: 10,
            itemsBeforeCollapse: 1,
        });

        expect(data.map((d) => d.output.itemsBeforeCollapse)).not.toContain(2);

        expect(sortByPriority(data.slice(1)).map((d) => d.output.itemsAfterCollapse)).toEqual([7, 6, 5, 4, 3, 2, 1]);

        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having itemsAfterCollapse fixed", () => {
        const data = generateAutoCollapseParameters({
            size: 10,
            itemsAfterCollapse: 2,
        });

        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(1);
        expect(data.map((d) => d.output.itemsAfterCollapse)).not.toContain(3);

        expect(sortByPriority(data.slice(1)).map((d) => d.output.itemsBeforeCollapse)).toEqual([6, 5, 4, 3, 2, 1]);

        expect(sortByPriority(data)).toMatchSnapshot();

    });

    it("generates parameters when having itemsBeforeCollapse set to a range", () => {
        const data = generateAutoCollapseParameters({
            size: 10,
            itemsBeforeCollapse: {
                min: 1,
                max: 2,
            },
        });

        expect(data.slice(1).map((d) => d.output.itemsBeforeCollapse)).not.toContain(3);
        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having itemsAfterCollapse set to a range", () => {
        const data = generateAutoCollapseParameters({
            size: 10,
            itemsAfterCollapse: {
                min: 2,
                max: 4,
            },
        });

        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(1);
        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(5);
        expect(sortByPriority(data)).toMatchSnapshot();
    });

    it("generates parameters when having both itemsBeforeCollapse and itemsAfterCollapse set to a range", () => {
        const data = generateAutoCollapseParameters({
            size: 10,
            itemsBeforeCollapse: {
                min: 1,
                max: 2,
            },
            itemsAfterCollapse: {
                min: 1,
                max: 2,
            },
        });
        expect(data.slice(1).map((d) => d.output.itemsBeforeCollapse)).toContain(2);
        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).toContain(2);
        expect(data.slice(1).map((d) => d.output.itemsBeforeCollapse)).not.toContain(3);
        expect(data.slice(1).map((d) => d.output.itemsAfterCollapse)).not.toContain(3);
        expect(sortByPriority(data)).toMatchSnapshot();
    });
});
