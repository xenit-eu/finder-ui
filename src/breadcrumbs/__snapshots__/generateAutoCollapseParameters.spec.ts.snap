// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breadcrumbs/generateAutoCollapseParameters generates one set of parameters when maxItems is set 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 4,
    },
    "priority": 1,
  },
]
`;

exports[`breadcrumbs/generateAutoCollapseParameters generates parameters when having both itemsBeforeCollapse and itemsAfterCollapse set to a range 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 2,
  },
]
`;

exports[`breadcrumbs/generateAutoCollapseParameters generates parameters when having itemsAfterCollapse fixed 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 6,
      "maxItems": 9,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 5,
      "maxItems": 9,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 4,
      "maxItems": 9,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 3,
      "maxItems": 9,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 3,
  },
]
`;

exports[`breadcrumbs/generateAutoCollapseParameters generates parameters when having itemsAfterCollapse set to a range 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 5,
      "maxItems": 9,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 4,
      "maxItems": 9,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 3,
      "maxItems": 9,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 3,
  },
]
`;

exports[`breadcrumbs/generateAutoCollapseParameters generates parameters when having itemsBeforeCollapse fixed 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 7,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 6,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 5,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 2,
  },
]
`;

exports[`breadcrumbs/generateAutoCollapseParameters generates parameters when having itemsBeforeCollapse set to a range 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 6,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 5,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 2,
  },
]
`;

exports[`breadcrumbs/generateAutoCollapseParameters generates parameters when no constraints are set 1`] = `
Array [
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 10,
    },
    "priority": 11,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 4,
      "itemsBeforeCollapse": 4,
      "maxItems": 9,
    },
    "priority": 8,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 4,
      "maxItems": 9,
    },
    "priority": 7,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 3,
      "itemsBeforeCollapse": 3,
      "maxItems": 9,
    },
    "priority": 6,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 3,
      "maxItems": 9,
    },
    "priority": 5,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 2,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 4,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 2,
      "maxItems": 9,
    },
    "priority": 3,
  },
  Object {
    "output": Object {
      "itemsAfterCollapse": 1,
      "itemsBeforeCollapse": 1,
      "maxItems": 9,
    },
    "priority": 2,
  },
]
`;
